---
title: '16S biplots with marginal density plots and boxplots'
author: "Jamie McMurtrie"
date: "21/6/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, error = TRUE)
knitr::opts_knit$set(root.dir = "C:/Users/jm1106/Malawi_Tilapia_Microbiomes")
```

## Set-up
For the PCA biplot, I'm using `factoextra`, which I've found easiest (http://www.sthda.com/english/wiki/factoextra-r-package-easy-multivariate-data-analyses-and-elegant-visualization). 

```{r libraries, include=TRUE}
library(here); packageVersion("here") # make the relative paths work
library(tidyverse); packageVersion("tidyverse")
library(phyloseq); packageVersion("phyloseq")
library(gridExtra); packageVersion("gridExtra") # for layout of separate plots
library(RColorBrewer); packageVersion("RColorBrewer")
library(factoextra); packageVersion("factoextra") # for lovely PCA biplots
library(ggpubr); packageVersion("ggpubr") # for publication-quality ggplots
library(ggExtra); packageVersion("ggExtra") # for density plots around scatter plots
library(cowplot); packageVersion("cowplot") # for density plots
```


```{r seed and plot settings, include=TRUE}
theme_set(theme_bw())
set.seed(1234)

#colours for pond sites
cbPalette.ponds.geog <- c("1" = "#e6ab02", "2" = "#66a61e", "3" = "#1b9e77","4" = "#a6761d", "5" = "#d95f02", "6" = "#7570b3", "7" = "#e7298a")
theme_set(theme(panel.background = element_blank(), axis.line = element_line(colour = "black")))
#when plotting with ggplot, removes grid lines etc for a clean plot

annoCol.pond <- list(Pond=c("1" = "#e6ab02", "2" = "#66a61e", "3" = "#1b9e77","4" = "#a6761d", "5" = "#d95f02", "6" = "#7570b3", "7" = "#e7298a"))
annoCol.sampleType <- list(sample_type=c("Pond" = "#0072B2", "Swab" = "#CC79A7"))
#pheatmap requires colour palletes as a list rather than named vector. 
#perhaps in future could call :
#annotation_colors = list(sample_type=cbPalette.sampleType)
```

A useful function for cleaning up phyloseq objects after subsetting to remove any zero taxa or samples, and optionally transform counts to relative abundance
```{r phyloseq tidy up functions}
tidyPS <- function(ps, RA){
  ps <- prune_taxa(taxa_sums(ps) > 0, ps)
  ps <- prune_samples(sample_sums(ps) > 0, ps)
  if(RA == TRUE){
    ps <- transform_sample_counts(ps, function(x){x / sum(x)})
  }
  return(ps)} 
```

## Tilapia pond water microbiomes, seperated by pond site
Load the 16S (V4) phyloseq object, subset samples , remove ASVs with fewer than 10 reads overall, and extract the various data tables (since I'm not going to use phyloseq beyond this step).  Removing the rare ASVs reduces the zero overinflation, though keeping these rare ASVs doesn't have much effect on the results. This appears particuarly the case for this Malawi dataset, and major site trends still hold true.
```{r load data PL tank plus direct inputs}
# 16S V4 data
V4.tree <- readRDS(here("data", "process", "V4_phyloseq_obj_with_tree_prev_filt.rds")) 

V4.tree #working with 5782 taxa by 67 samples

V4.sub <- subset_samples(V4.tree, sample_type %in% "Pond") %>% tidyPS(., FALSE)
sample_data(V4.sub)$Pond <- as.factor(sample_data(V4.sub)$Pond)
V4.sub #working with 5168 taxa by 35 samples

V4.pond <- prune_taxa(taxa_sums(V4.sub) > 10, V4.sub)
V4.pond ##working with 2403 taxa by 35 samples

asvs_V4 <- as.data.frame(otu_table(V4.pond))
asvs_V4 <- as_tibble(asvs_V4, rownames = "Sample_ID", colnames = "ASV_ID")

taxs_V4 <- as.data.frame(tax_table(V4.pond))
taxs_V4 <- as_tibble(taxs_V4, rownames = "ASV_ID")

samples_V4 <- as.data.frame(sample_data(V4.pond))
samples_V4 <- as_tibble(samples_V4, rownames = "SampleID")

# Make a streamlined metadata object, relastically we are only concerned by pond site in this case
sample_vars_V4 <- samples_V4 %>% select(SampleID, Pond, sample_type)
```

Carry out zero count replacement with `cmultRepl`, convert to proportions, then compute center mean log ratio transform. I then run the PCA (equivalent to PCoA with Euclidean distances), visualise the eigenvalues, and generate the biplot with the `factoextra` function `fviz_pca_biplot`.
```{r V4 biplot tank inputs, include =TRUE}
V4_czm <- zCompositions::cmultRepl(otu_table(V4.pond), method = "CZM") #zero replacment method, czm is  more roboust to zeros than gbm

# make our compositional dataset with the center mean log ratio function, i.e. log(x) - mean(log(x))
V4_comp <- CoDaSeq::codaSeq.clr(V4_czm) 

# Run the PCA on the CLR-transformed data
V4_PCA <- prcomp(V4_comp)

# Visualize eigenvalues/variances
fviz_screeplot(V4_PCA, addlabels = TRUE)


# Generate biplot with colours for compartments
pmain_V4 <- fviz_pca_biplot(V4_PCA, 
                axes = c(1, 2), # which PCs to plot
                repel = TRUE, # avoid text overlapping
                # Settings for samples
                geom = "point", 
                col.ind = sample_vars_V4$Pond, # Grouping variable, pond site
                pointsize = 4,
                pointshape = 19, alpha = 0.7,
                mean.point = FALSE, # Whether to include the mean of each group
                # Settings for ASVs (i.e. variables)
                geom.var = c("arrow", "text"), # Show arrows and text, no points
                select.var = list(contrib=15), # How many of the top variables to show
                col.var = "black") +
                #addEllipses=TRUE, ellipse.level=0.95) + #in this case as so tightly clustered eclipses don't enhance
   coord_fixed(1) + # run first with this command to get the best coordinates given the data
  #coord_cartesian(xlim = c(-60, 60), ylim = c(-30, 40)) + # specify coordinates to align boxplots
  scale_colour_manual(name = "Pond", values = cbPalette.ponds.geog) + # Custom colours
  labs(title = "")

pmain_V4
```




## Stats testing
Statistical testing of significant difference between pond sites. We utilise to tests here, firstly. a permanova through the adonis function that is the main test for difference between groups. (Note the test take the centroid i.e. median position of each group and uses this for the test) The seperation is so obvious by eye that we don't really need to do this, but if you want a P value here it is!
Additionally, we perform a permutation test for homgeneity of multivariate dispersions. As mentioned above the permanova just tests the centroid positioning, but its useful to include this second test to check if the dispersion of points within each group significanly differs. In this case its a no, as each group is tightly clustered. 
```{r stats, include=TRUE, echo=FALSE}

beta_permanova <- function(asvs, variableTest){
  set.seed(1234)
  clr_dist <- dist(asvs, method = "euclidean") #calculate distance
  veg_ad <- vegan::adonis(clr_dist ~ variableTest) #permanova test
  dispr <- vegan::betadisper(clr_dist, variableTest)
  print(veg_ad)
  print(dispr)
  print(vegan::permutest(dispr, pairwise = TRUE)) # still significant when we test for dispertion
  plot(dispr, main = "Ordination Centroids and Dispersion Labeled: Aitchison Distance", sub = "")
  #coef <- coefficients(veg_ad)["BNG_Inc",]
  #top.coef <- coef[rev(order(abs(coef)))[1:20]]
  
}
beta_permanova(V4_comp, sample_vars_V4$Pond)

```

## Heatmap of contributing ASVs
I'm struggling to find an elegant computational way to extract the top 15 contributing ASVs to a vector from the factor extra PCA object, so doing it manually. Want to plot as a heatmap to visualise abundance between samples and ponds. 

```{r top asvs heatmap tk2, include=TRUE, echo=FALSE}
library(microbiomeutilities)
contrib_asv <- c("ASV23", "ASV3", "ASV111", "ASV44", "ASV1", "ASV34", "ASV7", "ASV4", "ASV36", "ASV9", "ASV85", "ASV15", "ASV25", "ASV22", "ASV21") #manual curation of a vector for top 15 asvs contributing to dissimilarity between groups on PCA.
ps_clr <- V4.pond
otu_table(ps_clr) <- otu_table(V4_comp, taxa_are_rows = FALSE) # create a phyloseq object with compositional transformed counts
ps_contrib_asv <- subset_taxa(ps_clr, rownames(tax_table(ps_clr)) %in% contrib_asv) # subset to keep only the top 15 taxa
ps_contrib_asv

sample_data(ps_contrib_asv)$Pond <- as.factor(sample_data(ps_contrib_asv)$Pond)


custom_heatmap <- function(ps, VariableA, VariableB, core_tx){ #this function is lifted from microbiomeutilities::plot_taxa_heatmap, but bypassess their inforced transformations
    otu.mat <- abundances(ps)
    otu.mat <- otu.mat[match(core_tx, rownames(otu.mat)),] #reorder otu table to match allocation of core otus (i.e. not numeric ascending)
    #the swab samples are also in the incorrect order, here we extract the sample data and reorder by pond site, return just the sample names as a charatcer vector
    sample_order <- sample_data(ps) %>% as.data.frame() %>% .[order(.$Pond),] %>% row.names()
    otu.mat <- otu.mat[,match(sample_order, colnames(otu.mat))] #and reorder the otu table based on this
  meta.tab <- meta(ps)
  
  # choose which variables of interest to include in
  # the heatmap
  select.meta.A <- subset(meta.tab, select = c(VariableA))
  #select.meta.B <- subset(meta.tab, select = c(VariableB))
  #col_df <- data.frame(select.meta.A, select.meta.B)
  
  newnames <- vector(mode="character")
  tax_labels <- tax_table(ps)
  #tax_labels <- tax_labels[match(s16_all_core, rownames(tax_labels)),]
  #newnames <- paste(rownames(tax_labels),tax_labels[,"Class"])
  #newnames <- lapply(newnames,
  #  function(x) bquote(italic(.(x))))
  
  #for loop, to create a vector of the highest taxonomic classification avalabile to each ASV,for use as a label in plot
m <- 1
  highest_name <- vector(mode="character")
  for(x in rownames(tax_labels)){
    if(!is.na(tax_labels[x,"Genus"] %>% as.character())){
      highest_name[m] <- tax_labels[x,"Genus"] %>% as.character()
    }
    else if(!is.na(tax_labels[x,"Family"] %>% as.character())){
      highest_name[m] <- tax_labels[x,"Family"] %>% as.character() %>% paste(., "_F") %>% gsub(" ", "", .)
    }
    else if(!is.na(tax_labels[x,"Order"] %>% as.character())){
      highest_name[m] <- tax_labels[x,"Order"] %>% as.character() %>% paste(., "_O") %>% gsub(" ", "", .)
    }
    else if(!is.na(tax_labels[x,"Class"] %>% as.character())){
      highest_name[m] <- tax_labels[x,"Class"] %>% as.character() %>% paste(., "_C") %>% gsub(" ", "", .)
    }
    else if(!is.na(tax_labels[x,"Phylum"] %>% as.character())){
      highest_name[m] <- tax_labels[x,"Phylum"] %>% as.character() %>% paste(., "_P") %>% gsub(" ", "", .)
    }
    else if(!is.na(tax_labels[x,"Kingdom"] %>% as.character())){
      highest_name[m] <- tax_labels[x,"Kingdom"] %>% as.character() %>% paste(., "_K") %>% gsub(" ", "", .)
    }
    else{print("Error, undefined taxonomy")}
    newnames[m] <- paste(x, ":", highest_name[m]) #add the new labels as a vector, with highest tax classifcation and asv name.
    m <- m + 1
  }
  
  #reorder otumat by rownames of tax label so new labels are assigned correctly
  reorder.otu.mat <- otu.mat[match(rownames(tax_labels), rownames(otu.mat)),]
  rownames(reorder.otu.mat) <- newnames #assign new ASV labels as rownmaes of df for plotting
  
  #phylo_tree <- ape::chronos(phy_tree(ps))
  #phylo_hclust <- ape::as.hclust.phylo(phylo_tree)
  
  heatmap <- pheatmap::pheatmap(reorder.otu.mat,
                                #cluster_rows = phylo_hclust,
                                cluster_rows = TRUE,
                                cluster_cols = FALSE,
                                #labels_row = as.expression(newnames),
                                show_rownames = TRUE,
                                annotation_col = select.meta.A,
                                annotation_colors = annoCol.pond,
                                #annotation_row = row_df,
                                color = viridis::viridis(10))
  
  
  return(heatmap)
}
contrib_heatmap <- custom_heatmap(ps_contrib_asv, "Pond", "sample_type", contrib_asv)

ggsave(here("results", "figures", "pond_16s_ordination.svg"), plot = pmain_V4, width = 6, height = 5, dpi = 600)

#this is a better method to save, utilise cowplot to convert heatmap to a gg object, then can save as svg with ggsave. 
final_heatmap <- cowplot::plot_grid(contrib_heatmap$gtable, ncol = 1)
ggsave(here("results", "figures", "contrib_heatmap_pond_16s.svg"), plot = final_heatmap, width = 8, height = 5, dpi = 600)
```
N.b. ASV23 NA is order Kapabacteriales, add this label mannuall in inkscape, ran out of time to tinker. 

